yt_dlp/extractor/lastfm.py

```diff
@@ -0,0 +1,153 @@
+        tbody = self._search_regex(
+            r'(?s)tbody\s+data-playlisting-add-entries[^>]*>(.+?)</tbody>', webpage, 'tbody')
+        return [
+            self.url_result(player_url, 'Youtube')
+            for player_url in re.findall(r'href="([^"]+youtube.com/watch[^"]+)"', tbody)
+        ]
+
+    def _download_page(self, url, current_page_number, last_page_number, playlist_id):
+        note = 'Downloading page'
+        query = {}
+
+        if last_page_number:
+            note = '%s %d of %d' % (note, current_page_number, last_page_number)
+            query['page'] = current_page_number
+
+        return self._download_webpage(url, playlist_id, note, query=query)
+
+    def _entries(self, url, playlist_id):
+        webpage = self._download_webpage(url, playlist_id)
+        last_page_number = self._extract_last_page(webpage)
+        start_page_number = self._extract_current_page(url) or 1
+
+        for page_number in range(start_page_number, (last_page_number or start_page_number) + 1):
+            webpage = self._download_page(url, page_number, last_page_number, playlist_id)
+            page_entries = self._extract_entries(webpage)
+
+            for e in page_entries:
+                yield e
+
+    def _real_extract(self, url):
+        playlist_id = self._match_id(url)
+        return self.playlist_result(self._entries(url, playlist_id), playlist_id)
+
+
+class LastFMAlbumIE(LastFMPlaylistBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?last\.fm/music/(?P<artist_id>[^/]+)/(?P<id>[^+/#?][^/#?]+)(/|[^/]*/?)$'
+    _TESTS = [{
+        'url': 'https://www.last.fm/music/Oasis/(What%27s+the+Story)+Morning+Glory%3F',
+        'info_dict': {
+            'id': '(What%27s+the+Story)+Morning+Glory%3F',
+        },
+        'playlist_count': 12,
+    }]
+
+
+class LastFMPlaylistIE(LastFMPlaylistBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?last\.fm/(music|tag)/(?P<id>[^/#?]+)(/|/\+?tracks/?|[^/]*/?)$'
+    _TESTS = [{
+        'url': 'https://www.last.fm/music/Oasis',
+        'info_dict': {
+            'id': 'Oasis',
+        },
+        'playlist_count': 10,
+    }, {
+        'url': 'https://www.last.fm/music/Oasis/+tracks',
+        'only_matching': True,
+    }, {
+        'url': 'https://www.last.fm/tag/rock',
+        'only_matching': True,
+    }, {
+        'url': 'https://www.last.fm/tag/rock/tracks',
+        'only_matching': True,
+    }]
+
+
+class LastFMUserIE(LastFMPlaylistBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?last\.fm/user/(?P<user_id>[^/]+)/playlists/(?P<id>[^/#?]+)(/|[^/]*/?)$'
+    _TESTS = [{
+        'url': 'https://www.last.fm/user/mehq/playlists/12319471',
+        'info_dict': {
+            'id': '12319471',
+        },
+        'playlist_count': 30,
+    }]
+
+
+class LastFMIE(InfoExtractor):
+    _VALID_URL = r'''(?x)
+                    https?://(?:www\.)?last\.fm/
+                    music/(?P<artist_id>[^/]+)/(?P<album_id>[^/]+)/(?P<id>[^/#?]+)(/|[^/]*/?)$
+                  '''
```

Comment:
did u mean this?
```suggestion
    _VALID_URL = r'https?://(?:www\.)?last\.fm/music/[^/]+/[^/]+/(?P<id>[^/#?]+)'
```
I don't get what `(/|[^/]*/?)$` is trying to do